statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/feedback/FeedbackUnionOperator.java

```diff
@@ -84,14 +90,30 @@ public void processFeedback(T element) {
     if (closedOrDisposed) {
       return;
     }
-    if (isBarrierMessage.test(element)) {
-      feedbackLogger.commit();
+    OptionalLong maybeCheckpoint = isBarrierMessage.apply(element);
+    if (maybeCheckpoint.isPresent()) {
+      commitCheckpoint(maybeCheckpoint.getAsLong());
     } else {
       sendDownstream(element);
-      feedbackLogger.append(element);
+      appendToCheckpoint(element);
+    }
+  }
+
+  private void appendToCheckpoint(T element) {
+    for (UnboundedFeedbackLogger<T> logger : uncompletedCheckpoints.values()) {
+      logger.append(element);
     }
   }
 
+  private void commitCheckpoint(final long checkpointId) {
+    UnboundedFeedbackLogger<T> logger = uncompletedCheckpoints.remove(checkpointId);
```

Comment:
Maybe more of a question from my side:
Is this method guaranteed to be called for a given checkpoint id, even if the checkpoint already failed?
I'm wondering if that's a case where this `uncompletedCheckpoints` map grows indefinitely in size.

Reply:
It would eventually be called when a `checkpointId` succeed with a greater value.
You can check for a reference any of the `TwoPhaseCommit` operators that Flink has for a similar logic.
(for example StreamingFlieSink)